ts/private/ts_project_worker.js

```diff
@@ -540,23 +538,53 @@ function createProgram(args, initialInputs) {
     }
 }
 
+// How much (%) of memory should be free at all times. 
+const NEAR_OOM_ZONE = 20
+
+function isNearOomZone() {
+    const stat = v8.getHeapStatistics();
+    const used = (100 / stat.heap_size_limit) * stat.used_heap_size
+    return 100 - used < NEAR_OOM_ZONE
+}
+
+/** @type {Map<string, ReturnType<createProgram> & {previousInputs?: import("@bazel/worker").Inputs}>} */
+const workers = new Map();
+
+function sweepLeastRecentlyUsedWorkers() {
+    for (const [k, w] of workers) {
+        w.program.close();
+        workers.delete(k);
+        // stop killing workers as soon as the worker is out the oom zone 
+        if (!isNearOomZone()) {
```

Comment:
How about putting this at the start of the loop? Then we don't have to put it everywhere we invoke `sweepLeastRecentlyUsedWorkers`...

Reply:
yeah, that makes sense but way too smart thing to understand when we look at it later on. I am not sure.

Reply:
my thought process was `sweep workers if it is near the oom zone` and `kill workers until it gets out of the oom zone`

Reply:
But right now it essentially requires the `if` before invoking it..? Not a big deal though 🤷

Reply:
exactly. I find it easier to read this way.

Reply:
I was thinking just `avoid oom` and putting all the logic in one method. It's all internal though so doesn't really matter. Your choice 👍